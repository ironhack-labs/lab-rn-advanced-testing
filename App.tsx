import React, {useState} from 'react';
import {StyleSheet, Text, View, SafeAreaView} from 'react-native';
import Board from './src/components/board';
import {calculateWinner} from './src/utils';

const App = () => {
  const [history, setHistory] = useState([Array(9).fill(null)]);
  const [currentMove, setCurrentMove] = useState(0);
  const xIsNext = currentMove % 2 === 0;
  const currentSquares = history[currentMove];

  function handlePlay(nextSquares: string[]) {
    const nextHistory = [...history.slice(0, currentMove + 1), nextSquares];
    setHistory(nextHistory);
    setCurrentMove(nextHistory.length - 1);
  }

  const winner = calculateWinner(currentSquares);
  let status;
  if (winner) {
    status = 'Winner: ' + winner;
  } else {
    status = 'Next player: ' + (xIsNext ? 'X' : 'O');
  }

  return (
    <SafeAreaView>
      <View style={styles.container}>
        <Text testID="status" style={styles.status}>
          {status}
        </Text>
        <Board xIsNext={xIsNext} squares={currentSquares} onPlay={handlePlay} />
      </View>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {},
  status: {
    fontSize: 16,
    color: '#333',
    paddingVertical: 10,
  },
});

export default App;
