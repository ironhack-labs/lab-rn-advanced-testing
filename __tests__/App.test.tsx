import 'react-native';
import React from 'react';
import {render, fireEvent, waitFor} from '@testing-library/react-native';
import App from '../App';

// Note: import explicitly to use the types shiped with jest.
import {it} from '@jest/globals';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

describe('App', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Game Tic Tac Toe', () => {
  it('should render the game board with the expected UI elements', () => {
    const {getByTestId} = render(<App />);
    const board = getByTestId('board');
    const status = getByTestId('status');
    expect(board).toBeTruthy();
    expect(status).toBeTruthy();
    expect(status.props.children).toEqual('Next player: X');
  });

  it('should render rows of the game board', () => {
    const {getByTestId} = render(<App />);
    const board = getByTestId('board');
    expect(board).toBeTruthy();
    expect(board.children.length).toBe(3);
  });

  it('should allow users to tap on cells to make moves', async () => {
    const {getByTestId} = render(<App />);
    const square0 = getByTestId('square-0');
    fireEvent.press(square0);
    await waitFor(() => {
      expect(getByTestId('status').props.children).toBe('Next player: O');
    });
  });

  it('should update the game state when a move is made', async () => {
    const {getByTestId} = render(<App />);
    const square0 = getByTestId('square-0');
    fireEvent.press(square0);
    await waitFor(() => {
      expect(getByTestId('square-text-0').props.children).toBe('X');
    });
  });

  it('should detect a winner when the game is over', async () => {
    const {getByTestId} = render(<App />);
    const square0 = getByTestId('square-0');
    const square1 = getByTestId('square-1');
    const square4 = getByTestId('square-4');
    const square7 = getByTestId('square-7');
    const square8 = getByTestId('square-8');
    fireEvent.press(square0);
    fireEvent.press(square1);
    fireEvent.press(square4);
    fireEvent.press(square7);
    fireEvent.press(square8);

    await waitFor(() => {
      expect(getByTestId('status').props.children).toBe('Winner: X');
    });
  });
});
