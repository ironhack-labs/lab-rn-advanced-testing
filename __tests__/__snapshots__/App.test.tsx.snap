// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<RCTSafeAreaView>
  <View
    style={{}}
  >
    <Text
      style={
        {
          "color": "#333",
          "fontSize": 16,
          "paddingVertical": 10,
        }
      }
      testID="status"
    >
      Next player: X
    </Text>
    <View
      style={{}}
      testID="board"
    >
      <View
        style={
          {
            "flexDirection": "row",
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "borderColor": "#666",
              "borderWidth": 1,
              "height": 50,
              "justifyContent": "center",
              "margin": 2,
              "opacity": 1,
              "width": 50,
            }
          }
          testID="square-0"
        >
          <Text
            testID="square-text-0"
          />
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "borderColor": "#666",
              "borderWidth": 1,
              "height": 50,
              "justifyContent": "center",
              "margin": 2,
              "opacity": 1,
              "width": 50,
            }
          }
          testID="square-1"
        >
          <Text
            testID="square-text-1"
          />
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "borderColor": "#666",
              "borderWidth": 1,
              "height": 50,
              "justifyContent": "center",
              "margin": 2,
              "opacity": 1,
              "width": 50,
            }
          }
          testID="square-2"
        >
          <Text
            testID="square-text-2"
          />
        </View>
      </View>
      <View
        style={
          {
            "flexDirection": "row",
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "borderColor": "#666",
              "borderWidth": 1,
              "height": 50,
              "justifyContent": "center",
              "margin": 2,
              "opacity": 1,
              "width": 50,
            }
          }
          testID="square-3"
        >
          <Text
            testID="square-text-3"
          />
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "borderColor": "#666",
              "borderWidth": 1,
              "height": 50,
              "justifyContent": "center",
              "margin": 2,
              "opacity": 1,
              "width": 50,
            }
          }
          testID="square-4"
        >
          <Text
            testID="square-text-4"
          />
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "borderColor": "#666",
              "borderWidth": 1,
              "height": 50,
              "justifyContent": "center",
              "margin": 2,
              "opacity": 1,
              "width": 50,
            }
          }
          testID="square-5"
        >
          <Text
            testID="square-text-5"
          />
        </View>
      </View>
      <View
        style={
          {
            "flexDirection": "row",
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "borderColor": "#666",
              "borderWidth": 1,
              "height": 50,
              "justifyContent": "center",
              "margin": 2,
              "opacity": 1,
              "width": 50,
            }
          }
          testID="square-6"
        >
          <Text
            testID="square-text-6"
          />
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "borderColor": "#666",
              "borderWidth": 1,
              "height": 50,
              "justifyContent": "center",
              "margin": 2,
              "opacity": 1,
              "width": 50,
            }
          }
          testID="square-7"
        >
          <Text
            testID="square-text-7"
          />
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "borderColor": "#666",
              "borderWidth": 1,
              "height": 50,
              "justifyContent": "center",
              "margin": 2,
              "opacity": 1,
              "width": 50,
            }
          }
          testID="square-8"
        >
          <Text
            testID="square-text-8"
          />
        </View>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
